pkg/endpoint/endpoint.go

```diff
@@ -2403,3 +2433,61 @@ func (e *Endpoint) IPs() []net.IP {
 func (e *Endpoint) InsertEvent() {
 	e.getLogger().Info("New endpoint")
 }
+
+// must be called with endpoint mutex locked.
+func (e *Endpoint) syncPolicyMap() error {
+
+	if e.realizedMapState == nil {
+		e.realizedMapState = make(map[policymap.PolicyKey]struct{})
+	}
+
+	currentMapContents, err := e.PolicyMap.DumpToSlice()
+
+	if err != nil {
+		return fmt.Errorf("unable to dump PolicyMap for endpoint: %s", err)
+	}
+
+	for _, entry := range currentMapContents {
+		keyToAnalyze := policymap.PolicyKey{
+			Identity:         entry.Key.GetIdentity(),
+			DestPort:         byteorder.NetworkToHost(entry.Key.GetPort()).(uint16),
+			Nexthdr:          entry.Key.GetProto(),
+			TrafficDirection: entry.Key.GetDirection(),
+		}
+
+		// If key that is in policy map is not in desired state, just remove it.
+		if _, ok := e.desiredMapState[keyToAnalyze]; !ok {
+			err := e.PolicyMap.Delete(keyToAnalyze.Identity, keyToAnalyze.DestPort, u8proto.U8proto(keyToAnalyze.Nexthdr), policymap.TrafficDirection(keyToAnalyze.TrafficDirection))
+			if err != nil {
+				return err
+			}
+			// Remove from realized state
+			delete(e.realizedMapState, keyToAnalyze)
+		}
+	}
+
+	for keyToAdd := range e.desiredMapState {
+		if _, ok := e.realizedMapState[keyToAdd]; !ok {
+			err := e.PolicyMap.Allow(keyToAdd.Identity, keyToAdd.DestPort, u8proto.U8proto(keyToAdd.Nexthdr), policymap.TrafficDirection(keyToAdd.TrafficDirection))
```

Comment:
Another case where if we have access to the key, I don't see why we should need to know the details of the `policyKey` in this package, the whole key could be passed down.

Reply:
Agreed, I'll make a function that does this.